### ISSUE TITLE
post-install command not able to use proxy

### ISSUE THREAD

**Issue by dlpottra** [clearlinux/clr-installer#649]
**Describe the bug**
Post install script is not able to use proxy to install pip packages for a cloud image.

**To Reproduce**
1. clr-install -c cloud-native.yaml
cloud-native.yaml
...
post-install: [
   {cmd: "pip install collectd-gnocchi", chroot: true},
]
...

**Expected behavior**
python package should be installed.  

**Environment (please complete the following information):**
 - Clear Linux OS Version: 32230
 - Installer Version 2.3.7

**Additional context**
Proxies are defined in /etc/profile on the system running the installer.
Tried to use a bash script with proxies defined.
Tried to use the '--proxy' switch with pip in both a script and the post-install cmds.

**Comment by mdhorn** [comment]
@dlpottra  Please attach the actual YAML file used and the /root/clr-installer.log.

Commands run, for example as post install hooks, are logged along with the known environment variable in the clr-installer log. Do you see proxies in the list of cmd.Env?

As clr-installer always requires root, is your sudo environment set to pass proxy information?
```
$ sudo cat /etc/sudoers.d/visudo
Defaults        env_keep+="DISPLAY http_proxy https_proxy ftp_proxy no_proxy socks_proxy"
```

**Comment by dlpottra** [comment]
@mdhorn,

I do see the proxies in cmd.Env in the log file.  Here is the log file.
[clr-installer.log](https://github.com/clearlinux/clr-installer/files/4166132/clr-installer.log)

Here is fully enclosed script of what I am trying to do. 
[images.clear.sh.txt](https://github.com/clearlinux/clr-installer/files/4166117/images.clear.sh.txt)

I did not have /etc/sudoers.d/visudo.  I created it, but it didn't help.  Not sure if it is permissions thing.
drwxr-xr-x 2 root root   4.0K Feb  6 08:03 sudoers.d

-rw-r--r-- 1 root root 90 Feb  6 08:03 visudo

I also changed visudo to proxy.  That didn't help either.

**Comment by mdhorn** [comment]
@dlpottra Okay, the problem here is that networking is not fully configured on the target image, so when you chroot it doesn't work. Once the target boots, it would get it's networking from DHCP configurations -- forcing values at install would be a bad direction. The proxy values are there, but the resolver is not configured.

Instead of using chroot here, use the `--root <path` feature of pip. Also, you might want to include the --user flag to ensure that the content you install is not overwritten or remove by swupd.

Here is a good swag, untested of the changes you should consider:
```
--- images.clear.sh.txt	2020-02-06 09:52:38.000000000 -0800
+++ images.clear.sh.new	2020-02-06 11:28:28.000000000 -0800
@@ -1,6 +1,8 @@
 #!/bin/bash -ex
 set -o pipefail

+CHROOTPATH=$1
+
 # make directory for images
 mkdir glance-images || :

@@ -12,10 +14,10 @@
 #!/bin/bash -ex

 # add collectd-gnocchi
-pip install collectd-gnocchi
+pip install --root ${CHROOTPATH} collectd-gnocchi

 # tell collectd what to collect and where to put it.
-cat > /etc/collectd.conf << EOF2
+cat > ${CHROOTPATH}/etc/collectd.conf << EOF2
 LoadPlugin syslog
 LoadPlugin cpu
 LoadPlugin csv
@@ -75,7 +77,7 @@
 EOF2

 # create the collectd.service file
-cat > etc/systemd/system/collectd.service << EOF2
+cat > ${CHROOTPATH}/etc/systemd/system/collectd.service << EOF2
 [Unit]
 Description=Statistics collection and monitoring daemon
 After=local-fs.target network.target
@@ -149,8 +151,7 @@
 kernel: kernel-native

 post-install: [
-   {cmd: "cp \${yamlDir}/collectd.setup.sh \${chrootDir}/"},
-   {cmd: "\${yamlDir}/collectd.setup.sh", chroot: true},
+   {cmd: "\${yamlDir}/collectd.setup.sh \${chrootDir}"},
    {cmd: "systemctl enable collectd.service", chroot: true},
    {cmd: "rm -rf \${chrootDir}/tmp/"},
 ]

```

I'll update the [InstallerYAMLSyntax.md](https://github.com/clearlinux/clr-installer/blob/master/scripts/InstallerYAMLSyntax.md) with a note about commands requiring network access should not be used with 'chroot'.

**Comment by dlpottra** [comment]
Hi @mdhorn,

I have made the changes that you have suggested.  pip is still not installing the packages correctly.  pip is finding the packages on the machine and not installing the packages in the image.  I did see something were pip is looking at $HOME.  Changing that to the ${chrootDir}/root/ didn't help either.  Might be something else there that should be set.  Any suggestions?

Here is the failed part of the log:  (Should not see 'Requirement already satisfied' for collectd-gnocchi)
2020/02/06 12:19:21 [DBG] + pip install --root /tmp/install-575390791 collectd-gnocchi
2020/02/06 12:19:21 [DBG] Requirement already satisfied: collectd-gnocchi in /usr/lib/python3.8/site-packages (1.7.1)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: gnocchiclient>=4.0.0 in /usr/lib/python3.8/site-packages (from collectd-gnocchi) (7.0.5)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: keystoneauth1 in /usr/lib/python3.8/site-packages (from collectd-gnocchi) (3.18.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: debtcollector in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.22.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: monotonic in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.5)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: six in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.14.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: futurist in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.10.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: iso8601 in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (0.1.12)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: pbr>=1.4 in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (5.4.4)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: python-dateutil in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (2.8.1)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: cliff>=2.10 in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (2.18.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: ujson in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.35)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: requests>=2.14.2 in /usr/lib/python3.8/site-packages (from keystoneauth1->collectd-gnocchi) (2.22.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: os-service-types>=1.2.0 in /usr/lib/python3.8/site-packages (from keystoneauth1->collectd-gnocchi) (1.7.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: stevedore>=1.20.0 in /usr/lib/python3.8/site-packages (from keystoneauth1->collectd-gnocchi) (1.31.0)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: wrapt>=1.7.0 in /usr/lib/python3.8/site-packages (from debtcollector->gnocchiclient>=4.0.0->collectd-gnocchi) (1.11.2)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: PrettyTable<0.8,>=0.7.1 in /usr/lib/python3.8/site-packages (from futurist->gnocchiclient>=4.0.0->collectd-gnocchi) (0.7.2)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: pyparsing>=2.1.0 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (2.4.6)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: PyYAML>=3.12 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (5.3)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: cmd2!=0.8.3,<0.9.0,>=0.8.0 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (0.8.9)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (3.0.4)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: idna<2.9,>=2.5 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (2.8)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (1.25.8)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (2019.11.28)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: wcwidth; sys_platform != "win32" in /usr/lib/python3.8/site-packages (from cmd2!=0.8.3,<0.9.0,>=0.8.0->cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (0.1.8)
2020/02/06 12:19:21 [DBG] Requirement already satisfied: pyperclip in /usr/lib/python3.8/site-packages (from cmd2!=0.8.3,<0.9.0,>=0.8.0->cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (1.7.0)

**Comment by mdhorn** [comment]
What about using `--prefix` instead of `--root`?
Here the output looks different:
```
Collecting collectd-gnocchi
  Using cached https://files.pythonhosted.org/packages/7b/12/3c4c8c5c8a8c3be45e8a552bf494152924685fba1cdb252978b0ddf5f921/collectd_gnocchi-1.7.1-py2.py3-none-any.whl
Collecting gnocchiclient>=4.0.0
  Using cached https://files.pythonhosted.org/packages/3f/f8/b8bdd016fcefdd71f0138d9c5c360d7ff340940d4b9c7cb27b7dfbbeaf81/gnocchiclient-7.0.5-py2.py3-none-any.whl
Collecting keystoneauth1
  Using cached https://files.pythonhosted.org/packages/97/f6/6d17bcad6eff1561e3cfde1ea4d1d6bb9aac612b3fdfe78ec5049d01694e/keystoneauth1-3.18.0-py2.py3-none-any.whl
Requirement already satisfied: pbr>=1.4 in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (5.4.4)
Collecting cliff>=2.10
  Using cached https://files.pythonhosted.org/packages/55/2f/f1484035545a36de978c686b90a6d81c0b09a8e3002faf761737eef54cdc/cliff-2.18.0-py2.py3-none-any.whl
Collecting monotonic
  Using cached https://files.pythonhosted.org/packages/ac/aa/063eca6a416f397bd99552c534c6d11d57f58f2e94c14780f3bbf818c4cf/monotonic-1.5-py2.py3-none-any.whl
Requirement already satisfied: python-dateutil in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (2.7.5)
Requirement already satisfied: iso8601 in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (0.1.12)
Collecting futurist
  Using cached https://files.pythonhosted.org/packages/6e/56/cc28af95883fad946cc394fa4b326c2890bd392bd3b79ec9e22f6308d840/futurist-1.10.0-py2.py3-none-any.whl
Processing /root/.cache/pip/wheels/28/77/e4/0311145b9c2e2f01470e744855131f9e34d6919687550f87d1/ujson-1.35-cp38-cp38-linux_x86_64.whl
Collecting debtcollector
  Using cached https://files.pythonhosted.org/packages/bf/f5/0591d1fd42a546e543cff78d264511b63a176d2b406a94c6bec111cf25fa/debtcollector-2.0.0-py3-none-any.whl
Requirement already satisfied: six in /usr/lib/python3.8/site-packages (from gnocchiclient>=4.0.0->collectd-gnocchi) (1.13.0)
Collecting os-service-types>=1.2.0
  Using cached https://files.pythonhosted.org/packages/10/2d/318b2b631f68e0fc221ba8f45d163bf810cdb795cf242fe85ad3e5d45639/os_service_types-1.7.0-py2.py3-none-any.whl
Collecting stevedore>=1.20.0
  Using cached https://files.pythonhosted.org/packages/b1/e1/f5ddbd83f60b03f522f173c03e406c1bff8343f0232a292ac96aa633b47a/stevedore-1.31.0-py2.py3-none-any.whl
Requirement already satisfied: requests>=2.14.2 in /usr/lib/python3.8/site-packages (from keystoneauth1->collectd-gnocchi) (2.22.0)
Collecting cmd2!=0.8.3,<0.9.0,>=0.8.0
  Using cached https://files.pythonhosted.org/packages/e9/40/a71caa2aaff10c73612a7106e2d35f693e85b8cf6e37ab0774274bca3cf9/cmd2-0.8.9-py2.py3-none-any.whl
Requirement already satisfied: pyparsing>=2.1.0 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (2.4.6)
Requirement already satisfied: PrettyTable<0.8,>=0.7.2 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (0.7.2)
Requirement already satisfied: PyYAML>=3.12 in /usr/lib/python3.8/site-packages (from cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (5.1.2)
Requirement already satisfied: wrapt>=1.7.0 in /usr/lib/python3.8/site-packages (from debtcollector->gnocchiclient>=4.0.0->collectd-gnocchi) (1.11.2)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (3.0.4)
Requirement already satisfied: idna<2.9,>=2.5 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (2.8)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (1.25.7)
Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3.8/site-packages (from requests>=2.14.2->keystoneauth1->collectd-gnocchi) (2019.11.28)
Requirement already satisfied: pyperclip in /usr/lib/python3.8/site-packages (from cmd2!=0.8.3,<0.9.0,>=0.8.0->cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (1.7.0)
Requirement already satisfied: wcwidth; sys_platform != "win32" in /usr/lib/python3.8/site-packages (from cmd2!=0.8.3,<0.9.0,>=0.8.0->cliff>=2.10->gnocchiclient>=4.0.0->collectd-gnocchi) (0.1.7)
Installing collected packages: cmd2, stevedore, cliff, monotonic, futurist, ujson, debtcollector, os-service-types, keystoneauth1, gnocchiclient, collectd-gnocchi
Successfully installed cliff-2.18.0 cmd2-0.8.9 collectd-gnocchi-1.7.1 debtcollector-2.0.0 futurist-1.10.0 gnocchiclient-7.0.5 keystoneauth1-3.18.0 monotonic-1.5 os-service-types-1.7.0 stevedore-1.31.0 ujson-1.35
```

I'm not a python or pip expert. If that doesn't resolve the issue for you, I think I'll move this over to the clearlinux/distribution queue where we might get a broader audience that could help with the right pip command.

**Comment by dlpottra** [comment]
It didn't work for me.  I think you just installed a bunch of pip modules on your machine.  If you run it a second time, I think you will see all requirements are satisfied.

I do have a workaround.  I am downloading all the modules locally, copying them to the system and then doing a chroot pip install pointing to the local directory.

It would be nice to have this working.  It works with other image builders for other distro.  I think they enable the network and then clean it up before finalizing the image.

Up to you where you want to take this.

**Comment by mdhorn** [comment]
Fixed in release [2.4.0](https://github.com/clearlinux/clr-installer/releases/tag/2.4.0)

**Comment by mdhorn** [comment]
Released in Clear Linux OS 32380
